kafka-rest/src/main/java/io/confluent/kafkarest/requestlog/CustomLog.java

```diff
@@ -63,6 +69,44 @@ protected void doStop() throws Exception {
     }
   }
 
+  /**
+   * This class aggregates error-codes for produce-records within a single (http)produce-request.
+   * This implements toString() method which is used by CustomLog to get a message to log for the
+   * aggregated error counts.
+   */
+  public static class ProduceRecordErrorCounter {
+    private final Map<Integer, Integer> produceErrorCodeCountMap = new TreeMap<>();
+
+    public synchronized void incrementErrorCount(int httpErrorCode) {
+      produceErrorCodeCountMap.merge(httpErrorCode, 1, Integer::sum);
+    }
+
+    @Override
+    public String toString() {
+      return PRODUCE_ERROR_CODE_LOG_PREFIX
+          + produceErrorCodeCountMap.entrySet().stream()
+              .map(entry -> entry.getKey() + ":" + entry.getValue())
+              .collect(Collectors.joining(","));
+    }
+
+    @Override
+    public boolean equals(Object other) {
+      if (this == other) {
+        return true;
+      }
+      if (other == null || getClass() != other.getClass()) {
+        return false;
+      }
+      ProduceRecordErrorCounter that = (ProduceRecordErrorCounter) other;
+      return Objects.equals(produceErrorCodeCountMap, that.produceErrorCodeCountMap);
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hash(produceErrorCodeCountMap);
+    }
+  }
```

Comment:
While the issue might be benign, this construct contains races by the classical definition of a data race - two simultaneous accesses from different threads, where at least one access is a write, and there are no synchronization mechanisms between the accesses.
The writes are synchronized, so they are mutually exclusive with themselves, but there is no mechanism to guarantee either atomicity or visibility of the write actions to the read actions.

Furthermore, while defining `equals` and `hashCode` is generally good, here it might cause the wrong impression that this is usable as a map key - it's not really, as this is an inherently mutable type. In addition, the `Objects.hash` used in the `hashCode` implementation is a bit of a code smell - for a single element, it's actually just an obfuscated `Objects.hashCode()` (it shifts the instance hash code by 31).

Reply:
> there are no synchronization mechanisms between the accesses.

The data race between write and read of the map is unlikely. But such races are hard to track whenever they occur, so gr8 u called it out!

@wcheng92 can you make the `toString()` method synchronized?

> here it might cause the wrong impression that this is usable as a map key

A good call out again. Right the equals and hashCode are misleading.

@wcheng92 you should remove 2 methods equals() and hashCode(). But u would still need a way for easymock to compare 2 objects  `ProduceRecordErrorCounter` are logically equal(depending upon the maps contained within) Vs object equality. For this you can create a [custom-matcher](ProduceRecordErrorCounter) in easymock, which would do equality similar to equals() method.

Reply:
Thanks for helping to review the PR! 

Here are my thoughts about addressing the potential issues you mentioned:

1. Race condition issue: Synchronize `toString`, `equals`, and `hashCode` methods to prevent data race
2. Wrong impression for `equals` and `hashCode`: Make clearer comments for the class and mention this class **should NOT** be used as a map key
3. hashCode code smell: Remove `hashCode`

Please let me know if these potential solutions work or if there's a better solution ! Thank you so much!

Reply:
If we were to run a tool like spotbugs here, it would say that all variables that sit within a synchronized block, should sit within a synchronized block throughout the class, so it's the right thing to lock the toString here too.  

In reality, I actually don't think we care too much if the toString is slightly lagging in being correct though.  And I think it's only called when the overall produce request completes, ie no more writes are coming in, so the risk seems minimal.  And it's shouldn't concurrent modification exception either, as it's a read.  But, best to be correct and your code change looks good to me.

I see you've removed the hash and equals, thank you.